File: 288_Unique_Word_Abbreviation.py
_author_ = 'jake'
_project_ = 'leetcode'

# https://leetcode.com/problems/unique-word-abbreviation/
# An abbreviation of a word follows the form <first letter><number><last letter>.
# Assume you have a dictionary and given a word, find whether its abbreviation is unique in the dictionary.
# A word's abbreviation is unique if no other word from the dictionary has the same abbreviation.

# Create a set of words form dictionary to eliminate duplicates so the same word is not double counted. For each
# unique word, create a mapping from its abbreviation to the count of words with the same abbreviation. A word is
# unique if it is in the dictionary and no other word has the same abbreviation, or no other word has same abbreviation.
# Time - O(n) for init() where n is the total number of letters in all words in the dictionary. O(k) for isUnique()
# where k is the length of the word to be tested for uniqueness.
# Space - O(n)

from collections import defaultdict

class ValidWordAbbr(object):

    def __init__(self, dictionary):
        """
        :type dictionary: List[str]
        """
        self.dictionary = set(dictionary)
        self.freq = defaultdict(int)
        for word in self.dictionary:
            self.freq[self.abbreviate(word)] += 1

    def isUnique(self, word):
        """
        :type word: str
        :rtype: bool
        """
        abbr = self.abbreviate(word)
        if word in self.dictionary:
            return self.freq[abbr] == 1
        else:
            return abbr not in self.freq

    def abbreviate(self, word):
        n = len(word)
        if n < 3:
            return word
        return word[0] + str(n - 2) + word[-1]

File: java/288_Unique_Word_Abbreviation.java
/*
https://leetcode.com/problems/unique-word-abbreviation/
An abbreviation of a word follows the form <first letter><number><last letter>.
Assume you have a dictionary and given a word, find whether its abbreviation is unique in the dictionary.
A word's abbreviation is unique if no other word from the dictionary has the same abbreviation.

Note that words of length less than 2 are abbreviated to themselves.
Create a map from abbreviations to words.  If an abbreviation has already been used then it is mapped to the empty
String.  New word has unique abbreviation if that abbreviation has not been used or has only been used to abbreviate
that word.
Time - O(n) for constructor, O(1) for isUnique()
Space - O(n)
*/



public class ValidWordAbbr {

    private Map<String, String> map;                // instance variable (adding static could make class variable)

    public ValidWordAbbr(String[] dictionary) {     // constructor
        map = new HashMap<String, String>();

        for (String word : dictionary) {
            String abbreviation = abbreviate(word);

            if (map.containsKey(abbreviation) && !map.get(abbreviation).equals(word))   // .equals() to compare Strings
                map.put(abbreviation, "");
            else
                map.put(abbreviation, word);
        }
    }

    public boolean isUnique(String word) {
        String abbreviation = abbreviate(word);
        return !map.containsKey(abbreviation) || map.get(abbreviation).equals(word);
    }

    public String abbreviate(String word) {
        if (word.length() <= 2)
            return word;
        return word.charAt(0) + Integer.toString(word.length()-2) + word.charAt(word.length()-1);
    }

}
// Your ValidWordAbbr object will be instantiated and called as such:
// ValidWordAbbr vwa = new ValidWordAbbr(dictionary);
// vwa.isUnique("Word");
// vwa.isUnique("anotherWord");
